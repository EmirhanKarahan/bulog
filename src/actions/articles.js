import moment from "moment";
import { history } from "../routers/AppRouter";
import {
  addArticleFirebase,
  deleteImageFirebase,
  editArticleFirebase,
  getArticleByIdFirebase,
  getArticlesFirebase,
  removeArticleFirebase,
  uploadImageFirebase,
} from "../firebase/utils";

export const addArticle = (article) => {
  return {
    type: "ADD_ARTICLE",
    article,
  };
};

export const startAddArticle = (articleData = {}) => {
  return async (dispatch) => {
    const {
      title = "",
      subtitle = "",
      content = "",
      image = undefined,
      date = moment().unix(),
    } = articleData;

    const imageUrl = await uploadImageFirebase(image);
    const article = { title, subtitle, content, date, imageUrl };
    const ref = await addArticleFirebase(article);
    dispatch(addArticle({ id: ref.key, ...article }));
    history.push("/");
  };
};

export const editArticle = (id, updates) => ({
  type: "EDIT_ARTICLE",
  id,
  updates,
});

export const startEditArticle = (
  id,
  { title, subtitle, content, image, imageUrl }
) => {
  return async (dispatch) => {
    if (image) {
      await deleteImageFirebase(imageUrl);
      imageUrl = await uploadImageFirebase(image);
    }
    const updates = { title, subtitle, content, imageUrl };
    await editArticleFirebase(id, updates);
    dispatch(editArticle(id, updates));
    history.push(`/read/${id}`);
  };
};

export const removeArticle = ({ id } = {}) => ({
  type: "REMOVE_ARTICLE",
  id,
});

export const startRemoveArticle = ({ id } = {}) => {
  return async (dispatch, getState) => {
    dispatch(removeArticle({ id }));
    const snapshot = await getArticleByIdFirebase(id);
    const imageUrl = snapshot.val().imageUrl;
    await deleteImageFirebase(imageUrl);
    await removeArticleFirebase(id);
  };
};

export const setArticles = (articles) => ({
  type: "SET_ARTICLES",
  articles,
});

export const startSetArticles = () => {
  return async (dispatch, getState) => {
    const snapshot = await getArticlesFirebase();
    const articles = [];
    snapshot.forEach((childSnapshot) => {
      articles.push({ id: childSnapshot.key, ...childSnapshot.val() });
    });
    dispatch(setArticles(articles));
  };
};
