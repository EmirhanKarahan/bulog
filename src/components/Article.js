import React from "react";
import { connect } from "react-redux";
import moment from "moment";
import { Redirect } from "react-router-dom";

const Article = ({ article }) => {
  return article ? (
    <article className="article">
      <div className="content-container">
        <div className="article__content">
          <img src={article.imageUrl} className="article__image" />
          <div className="article__meta">
            <h2 className="article__title">{article.title}</h2>
            <h4 className="article__subtitle">{article.subtitle}</h4>
            <div className="article__text">{article.content}</div>
            <span className="article__author">{article.author}</span>
            <span className="article__date">
              {moment.unix(article.date).format("MMMM Do, YYYY")}
            </span>
          </div>
        </div>
      </div>
    </article>
  ) : (
    <Redirect to="/404"></Redirect>
  );
};

const mapStateToProps = (state, props) => {
  return {
    article: state.articles.find(
      (article) => article.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps)(Article);
