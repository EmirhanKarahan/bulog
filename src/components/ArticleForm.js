import React from "react";
import { Form, Field, withFormik } from "formik";
import AutoSizeTextArea from "react-textarea-autosize";
import * as Yup from "yup";

function ArticleForm({ values, errors, touched, isSubmitting }) {
  return (
    <Form className="article-form">
      <div className="article-form__error-area">
        {Object.keys(touched).map((field) => {
          if (errors[field]) {
            return (
              <span key={field} className="error-msg">
                {errors[field]}
              </span>
            );
          }
        })}
      </div>
      <Field
        type="text"
        name="title"
        className="text-input text-input--title"
        placeholder="Title"
        autoFocus
      />
      <Field
        type="text"
        name="subtitle"
        className="text-input text-input--subtitle"
        placeholder="Subtitle"
      />
      <Field
        as={AutoSizeTextArea}
        name="content"
        className="text-input text-input--content"
        placeholder="Your unique blog"
      />
      <label className="file-input__label mb-1">
        {values.image ? values.image.split("\\").slice(-1) : "Choose an image"}
        <Field
          type="file"
          name="image"
          accept=".jpeg, .jpg, .jfif, .webp., .png"
          className="file-input"
        />
      </label>
      <button
        disabled={isSubmitting}
        type="submit"
        className="button button--secondary mb-2"
      >
        Submit
      </button>
    </Form>
  );
}

const ArticleFormik = withFormik({
  mapPropsToValues({ title, subtitle, content, image }) {
    return {
      title: title || "",
      subtitle: subtitle || "",
      content: content || "",
      image: image || "",
    };
  },
  validationSchema: Yup.object().shape({
    title: Yup.string().min(2).max(10).required(),
    subtitle: Yup.string().min(2).max(10).required(),
    content: Yup.string().min(2).required(),
  }),
  handleSubmit(values, { resetForm, setErrors, setSubmitting }) {
    console.log(values);
    setSubmitting(false);
    resetForm();
  },
})(ArticleForm);

export default ArticleFormik;
