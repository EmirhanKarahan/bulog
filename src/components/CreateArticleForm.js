import React from "react";
import { Form, Field, withFormik } from "formik";
import AutoSizeTextArea from "react-textarea-autosize";
import * as Yup from "yup";

export function CreateArticleForm({
  values,
  errors,
  touched,
  isSubmitting,
  setFieldValue,
}) {
  return (
    <Form className={"create-article-form"}>
      {isSubmitting && (
        <img className="spinner" src="/images/spinner.svg" alt="spinner" />
      )}
      <div className="error-area">
        {touched &&
          Object.keys(touched).map((field) => {
            if (errors[field]) {
              return (
                <span key={field} className="error-msg">
                  {errors[field]}
                </span>
              );
            }
          })}
      </div>
      <Field
        type="text"
        name="title"
        disabled={isSubmitting}
        className="text-input text-input--title"
        placeholder="Title"
        autoFocus
      />
      <Field
        type="text"
        name="subtitle"
        disabled={isSubmitting}
        className="text-input text-input--subtitle"
        placeholder="Subtitle"
      />
      <Field
        as={AutoSizeTextArea}
        name="content"
        disabled={isSubmitting}
        className="text-input text-input--content"
        placeholder="Your unique blog post"
      />
      <label className="file-input__label mb-1">
        {values?.image ? values.image.name : "Choose an image"}
        <input
          type="file"
          name="image"
          disabled={isSubmitting}
          accept=".jpeg, .jpg, .jfif, .webp., .png"
          className="file-input"
          onChange={(event) => {
            setFieldValue("image", event.target.files[0]);
          }}
        />
      </label>
      <button
        disabled={isSubmitting}
        type="submit"
        className="button button--secondary mb-2"
      >
        Submit
      </button>
    </Form>
  );
}

const CreateArticleFormik = withFormik({
  mapPropsToValues({ title, subtitle, content, image }) {
    return {
      title: title || "",
      subtitle: subtitle || "",
      content: content || "",
      image: image || undefined,
    };
  },
  validationSchema: Yup.object().shape({
    title: Yup.string().min(10).max(80).required(),
    subtitle: Yup.string().min(15).max(100).required(),
    content: Yup.string().min(100).required(),
    image: Yup.mixed().required(),
  }),
  handleSubmit(values, { resetForm, props }) {
    props.onSubmit(values);
  },
})(CreateArticleForm);

export default CreateArticleFormik;
