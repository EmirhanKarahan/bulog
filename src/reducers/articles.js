const articlesReducerDefaultState = [];

export default (state = articlesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_ARTICLE':
      return [
        ...state,
        action.article
      ];
    case 'REMOVE_ARTICLE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_ARTICLE':
      return state.map((article) => {
        if (article.id === action.id) {
          return {
            ...article,
            ...action.updates
          };
        } else {
          return article;
        };
      });
    case 'SET_ARTICLES':
      return action.articles;
    default:
      return state;
  }
};