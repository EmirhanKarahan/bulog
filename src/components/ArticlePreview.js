import React from "react";
import moment from "moment";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { startRemoveArticle } from "../actions/articles";

function ArticlePreview({
  id,
  author,
  title,
  subtitle,
  date,
  imageUrl,
  editable,
  ...rest
}) {
  return (
    <article className="article-preview">
      <div className="content-container">
        <div className="article-preview__content">
          <div className="article-preview__meta">
            <Link className="url-link" to={`/read/${id}`}>
              <h2 className="article-preview__title">{title}</h2>
              <h4 className="article-preview__subtitle">{subtitle}</h4>
            </Link>

            <span className="article-preview__author">{author}</span>
            <span className="article-preview__date">
              {moment(date).format("MMMM Do, YYYY")}
            </span>
          </div>

          <div className="article-preview__right">
            <Link className="url-link" to={`/read/${id}`}>
              <img
                src={imageUrl}
                to={`/read/${id}`}
                className="article-preview__image"
              />
            </Link>

            {editable ? (
              <div className="article-preview__right__toolbox">
                <Link className="url-link" to={`/edit/${id}`}>
                  <img
                    src="/images/gear.svg"
                    height="20px"
                    width="20px"
                    alt=""
                  />
                  <b>Edit</b>
                </Link>
                <button
                  className="button button--danger"
                  onClick={() => {
                    rest.dispatch(startRemoveArticle({ id }));
                  }}
                >
                  Delete
                </button>
              </div>
            ) : (
              ""
            )}
          </div>
        </div>
      </div>
    </article>
  );
}

export default connect()(ArticlePreview);
