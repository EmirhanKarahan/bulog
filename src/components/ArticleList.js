import React from 'react'
import { connect } from 'react-redux';
import ArticlePreview from './ArticlePreview'

const ArticleList = ({articles}) => {
    return articles.map((article) => <ArticlePreview key={article.id} {...article} />)
}

const mapStateToProps = (state) => {
    return {
        articles: state.articles
    };
};

export default connect(mapStateToProps)(ArticleList)