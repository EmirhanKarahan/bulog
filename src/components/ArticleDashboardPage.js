import React from "react";
import ArticleList from "./ArticleList";

export default function ArticleDashboardPage() {
  return <ArticleList></ArticleList>;
}
