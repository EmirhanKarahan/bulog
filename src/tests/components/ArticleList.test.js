import React from "react";
import { shallow } from "enzyme";
import {ArticleList} from "../../components/ArticleList";
import articles from "../fixtures/articles"

test("should render ArticleList with articles", () => {
  const wrapper = shallow(<ArticleList articles={articles}></ArticleList>);
  expect(wrapper).toMatchSnapshot();
});

test("should render ArticeList with 'no article' message", ()=>{
    const wrapper = shallow(<ArticleList articles={[]}></ArticleList>);
    expect(wrapper).toMatchSnapshot();
})