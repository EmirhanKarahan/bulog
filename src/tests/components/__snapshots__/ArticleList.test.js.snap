// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ArticeList with 'no article' message 1`] = `
<div
  className="content-container"
>
  There is no article 😔 Please, add one 😁
</div>
`;

exports[`should render ArticleList with articles 1`] = `
Array [
  <Connect(ArticlePreview)
    content="Cool article"
    date={0}
    id="1"
    imageUrl="https://firebasestorage.googleapis.com/1"
    key="1"
    subtitle="First Article's Subtitle"
    title="First Article"
  />,
  <Connect(ArticlePreview)
    content="Really nice article"
    date={1000}
    id="2"
    imageUrl="https://firebasestorage.googleapis.com/2"
    key="2"
    subtitle="Second Article's Subtitle"
    title="Second Article"
  />,
  <Connect(ArticlePreview)
    content="Superb article"
    date={-1000}
    id="3"
    imageUrl="https://firebasestorage.googleapis.com/3"
    key="3"
    subtitle="Third Article's Subtitle"
    title="Third Article"
  />,
]
`;
